File: SundeeFundee/Domain/Calculations/PlateCalculation.swift
import Foundation

enum PlateCalculation {

    static let standardBarKg: Double = 20.0
    static let womenBarKg:    Double = 15.0
    static let availablePlatesKg: [Double] = [25, 20, 15, 10, 5, 2.5, 1.25]

    /// Returns the plates needed per side of the barbell.
    /// Keys are plate weights (kg), values are count per side.
    static func platesPerSide(
        totalWeightKg: Double,
        barbellWeightKg: Double = standardBarKg
    ) -> [(weight: Double, count: Int)] {
        guard totalWeightKg > barbellWeightKg else { return [] }

        var remaining = (totalWeightKg - barbellWeightKg) / 2.0
        var breakdown: [(weight: Double, count: Int)] = []

        for plate in availablePlatesKg {
            guard remaining >= plate else { continue }
            let count = Int(remaining / plate)
            if count > 0 {
                breakdown.append((weight: plate, count: count))
                remaining -= Double(count) * plate
            }
        }
        return breakdown
    }

    /// Human-readable description, e.g. "2×25kg + 1×5kg per side"
    static func description(totalWeightKg: Double, barbellWeightKg: Double = standardBarKg) -> String {
        let plates = platesPerSide(totalWeightKg: totalWeightKg, barbellWeightKg: barbellWeightKg)
        if plates.isEmpty {
            return "Bar only (\(Int(barbellWeightKg))kg)"
        }
        let parts = plates.map { p -> String in
            let w = p.weight.truncatingRemainder(dividingBy: 1) == 0
                  ? "\(Int(p.weight))kg"
                  : "\(p.weight)kg"
            return "\(p.count)×\(w)"
        }
        return parts.joined(separator: " + ") + " per side"
    }
}

File: web-app/src/lib/domain/plate-calculation.ts
export const STANDARD_BARBELL_KG = 20;
export const WOMENS_BARBELL_KG = 15;
export const AVAILABLE_PLATES = [25, 20, 15, 10, 5, 2.5, 1.25];

export interface PlateCount {
  weight: number;
  count: number;
}

/**
 * Returns the plates needed per side to reach the target total weight.
 * Uses a greedy algorithm with the standard plate sizes.
 *
 * @param totalWeightKg  The desired total barbell weight in kg.
 * @param barbellWeightKg  The weight of the barbell itself.
 * @returns Array of { weight, count } for each plate denomination used (per side).
 */
export function platesPerSide(totalWeightKg: number, barbellWeightKg: number): PlateCount[] {
  const perSide = (totalWeightKg - barbellWeightKg) / 2;
  if (perSide <= 0) return [];

  const result: PlateCount[] = [];
  let remaining = perSide;

  for (const plate of AVAILABLE_PLATES) {
    if (remaining <= 0) break;
    const count = Math.floor(remaining / plate);
    if (count > 0) {
      result.push({ weight: plate, count });
      remaining -= count * plate;
      // Prevent floating point drift
      remaining = Math.round(remaining * 1000) / 1000;
    }
  }

  return result;
}

/**
 * Returns a human-readable description of the plates needed per side.
 * Example: "2×20 + 1×5 + 1×2.5 per side"
 */
export function plateDescription(totalWeightKg: number, barbellWeightKg: number): string {
  const plates = platesPerSide(totalWeightKg, barbellWeightKg);
  if (plates.length === 0) return "Barbell only";
  const parts = plates.map((p) => `${p.count}\u00d7${p.weight}`);
  return `${parts.join(" + ")} per side`;
}
